### PR TITLE
fix(unstable_derive): trap atom methods

### PR DESCRIPTION
**Pull request by dai-shi**
This enables to trap `atom.read`, `atom.write` and `atom.onMount`.
This eliminates `originAtomState` for `getAtomState`.

### CODE REVIEW

File: src/vanilla/store.ts [hunk 1]
```diff
@@ -718,7 +729,7 @@ const buildStore = (getAtomState: StoreArgs[0]): Store => {
   }
 
   const unstable_derive = (fn: (...args: StoreArgs) => StoreArgs) =>
-    buildStore(...fn(getAtomState))
+    buildStore(...fn(getAtomState, atomRead, atomWrite))
```
Comment (dmaskasky) [inline src/vanilla/store.ts:732]: might not be necessary, not sure though.

File: src/vanilla/store.ts [hunk 2]
```diff
@@ -245,7 +251,12 @@ type Store = PrdStore | (PrdStore & DevStoreRev4)
 export type INTERNAL_DevStoreRev4 = DevStoreRev4
 export type INTERNAL_PrdStore = PrdStore
 
-const buildStore = (getAtomState: StoreArgs[0]): Store => {
+const buildStore = (
+  getAtomState: StoreArgs[0],
+  atomRead: StoreArgs[1],
+  atomWrite: StoreArgs[2],
+  atomOnMount: StoreArgs[3],
+): Store => {
```
Comment (dmaskasky) [inline src/vanilla/store.ts:259]:
```suggestion
const buildStore = (
  ...[getAtomState, atomRead, atomWrite, atomOnMount]: StoreArgs
): Store => {
```

**Comment by dmaskasky** [comment]
I like the atomOnMount trap.

I'm not sure about removing originalAtomState just yet though. It could be important for jotai-scope if it could be used to reliably resolve atomState. Earlier implementation was consistent for readAtomState and writeAtomState, but inconsistent / missing in other cases such as subscribe.

File: src/vanilla/store.ts (hunk 2 above)
Comment (dai-shi) [reply src/vanilla/store.ts:259]: nice!

**Comment by dai-shi** [comment]
> I'm not sure about removing originalAtomState just yet though.

As it's technically impossible to solve missing cases, I think it was an insufficient idea. Let's drop it and merge this.

**Comment by dmaskasky** [comment]
> As it's technically impossible to solve missing cases, I think it was an insufficient idea. Let's drop it and merge this.

Yes I agree.

**Pull request merged**